### ISSUE TITLE
Optionally log change of conditions inside noEvent()

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#9721]
Models with conditional expressions within `noEvent()`, possibly obtained by processing the `smooth()` operator, can cause solver issues. The macroscopic observed behaviour in some cases is that the solver takes a lot of very short steps, and possibly with a lot of Jacobian recomputations in the case of stiff solvers, around one per step or more. Debugging this kind of issues may be very challenging, as demonstrated, e.g., by #9131.

The root cause of the problem is apparently due to the fact that `noEvent()` is applied to discontinuous expressions, which may lead to sliding modes and chattering, or to expressions with discontinuous derivative, which can cause convergence problems in the solution of implicit integration formulae such as BDF, when the state trajectory gets in the neighbourhood of these discontinuities.

In particular, a discontinuous Jacobianthe seems to be the problem with #9131, as it was observed that one element of the Jacobian changes wildly every time the Jacobian is recomputed, even though the states are not practically changing, due to the very short time steps taken. If a symbolic Jacobian was computed, one would see two distinct values (the right and left limit), but in the case of #9131 a numerical Jacobian is computed, so the values fluctuate in an apparently random way.

Apparently, computing numerical Jacobians makes this problem even worse, since it was observed that turning on --generateSymbolicJacobians solves the issue in some cases.

It is still not entirely clear if these problems are caused by questionable Modelica code, or by some yet unknown issue with the runtime or code generation, or both. However, debugging them would be a lot easier if one knew what is the noEvent() expression(s) which is actually being computed across the discontinuity, when this issue arises.

This could be obtained by monitoring the Boolean conditions of conditional expressions inside noEvent(), if the appropriate flag, e.g. `-lv=LOG_NOEVENT` is turned on.

This would be similar to what is already done with `-lv=LOG_EVENT`, with some important differences
- noEvent expressions will still be evaluated literally, no actual event handling
- no need to generate zero-crossing functions
- changes in the condition would need to be monitored every time the RHS of the ODEs (or the residuals of the DAE) is computed, not only between accepted step

The last item is probably a bit tricky to implement, but is necessary. As the MWE in #9711 clearly demonstrates, accepted steps may happen to only lie on one side of the discontinuity, preventing its discovery if the conditions are only evaluated at the end of the step.

@AnHeuermann, @kabdelhak, I'm not sure how much work it would be to implement this logging feature, but I don't see any other way to fix BR's issues #9131 and Buildings issues #9256. What do you think?

**Comment by AnHeuermann** [comment]
> changes in the condition would need to be monitored every time the RHS of the ODEs (or the residuals of the DAE) is computed, not only between accepted steps

This is indeed hard to do in the compiler, because we loose the information that some expression `exp` inside `noEvent(exp)` comes from `noEvent` has to do extra stuff for logging. We could do something similar to zero-crossings though.

If `noEvent` is around a boolean we can add some logging of changed conditions. I did this manually for #9131, but with no relevant outputs.
If `exp` is of type real it get's a bit harder to log. We can have stuff like `noEvent(mod(y, 2.5))` or `noEvent(floor(y))`. Noticing a jump here is harder to do. The value for `mod` is expected to change with each step, but when is it a jump?

**Comment by casella** [comment]
> > changes in the condition would need to be monitored every time the RHS of the ODEs (or the residuals of the DAE) is computed, not only between accepted steps
> 
> This is indeed hard to do in the compiler, because we lose the information that some expression `exp` inside `noEvent(exp)` comes from `noEvent` has to do extra stuff for logging. We could do something similar to zero-crossings though.

Yes, that would be the idea. Of course it would have to be done on purpose.

> If `noEvent` is around a boolean we can add some logging of changed conditions.

As I understand from the specification, noEvent suppresses event generation by Real relations ([Section 3.7.5](url)) or by event-generating functions ([Section 3.7.2](https://specification.modelica.org/maint/3.5/operators-and-expressions.html#event-triggering-mathematical-functions)). There's no point putting it around a Boolean. 

> I did this manually for #9131, but with no relevant outputs.

Indeed, as the problem was apparently with delay(expr, 0) 😅 

> If `exp` is of type real it get's a bit harder to log. We can have stuff like `noEvent(mod(y, 2.5))` or `noEvent(floor(y))`. Noticing a jump here is harder to do. The value for `mod` is expected to change with each step, but when is it a jump?

I'm not sure I get your point. We should already have mechanisms in place to generate events for all the operators in [Section 3.7.2](https://specification.modelica.org/maint/3.5/operators-and-expressions.html#event-triggering-mathematical-functions), shouldn't we?

The point of my proposal would be to only retain the primary event-detection mechanism for logging even when these expressions are inside `noEvent`, of course without actually generating events, nor trying to find the exact point in time where the zero-crossing function changes the sign.

**Comment by AnHeuermann** [comment]
> There's no point putting it around a Boolean.

What I meant was uses like `noEvent(a>0)`, where the type of the expression `a>0` is Boolean. We want to know if this Boolean changes rapidly.

> I'm not sure I get your point. We should already have mechanisms in place to generate events for all the operators in [Section 3.7.2](https://specification.modelica.org/maint/3.5/operators-and-expressions.html#event-triggering-mathematical-functions), shouldn't we?

We probably can generate the same "zero-crossings" we generate for the event case (no `noEvent`) and with this check after each time step if they changed compared to the last step. Problem is, we can only call this after an ODE solver step. We won't catch changes in `noEvent` that happen if the right-hand side of the ODE is evaluated multiple times, like in https://github.com/OpenModelica/OpenModelica/issues/9711.

**Comment by casella** [comment]
> > There's no point putting it around a Boolean.
> 
> What I meant was uses like `noEvent(a>0)`, where the type of the expression `a>0` is Boolean. We want to know if this Boolean changes rapidly.

OK, now I see what you mean.
 
> > I'm not sure I get your point. We should already have mechanisms in place to generate events for all the operators in [Section 3.7.2](https://specification.modelica.org/maint/3.5/operators-and-expressions.html#event-triggering-mathematical-functions), shouldn't we?
> 
> We probably can generate the same "zero-crossings" we generate for the event case (no `noEvent`) and with this check after each time step if they changed compared to the last step. Problem is, we can only call this after an ODE solver step. We won't catch changes in `noEvent` that happen if the right-hand side of the ODE is evaluated multiple times, like in #9711.

Yes, this is the main issue.

**Comment by casella** [comment]
@AnHeuermann maybe I came up with a reasonable brute force solution, which is IMO good enough for debugging and should be straightforward to implement.

For each Boolean conditional expression that shows up in if-expressions enclosed by `noEvent`, the backend generates a Boolean `$noEventXXX` auxiliary variable and an equation assigning it to a that conditional expression with a noEvent. For example:
```modelica
z = noEvent(if condition1 and x > 0 and y > 0 then x + y else 0);
```
will generate one auxiliary variable `Boolean $noEvent1` and the following equation:
```modelica
$noEvent1 = noEvent(condition1 and x > 0 and y > 0);
```
This should be relatively straightforward, since we already have similar provisions in the backend for expressions that are _not_ enclosed by `noEvent`.

The runtime generates a `noEvent_log.csv` file, with one column for `time` and for each `$noEvent` variable. Every time the RHS of the ODE or the residuals of the DAE in daeMode is computed (for whatever reason), one line is added to this file.

In this way it is easy to catch chattering that happens within a single simulation step, e.g. because of R-K stages, Jacobian computations, iteration of implicit ODE/DAE solvers, etc, by simply looking at 0/1/0/1 adjacent changes in the CSV file.

What do you think?

This may be extremely helpful for debugging slow Buildings models, since apparently solving the `delayTime = 0` issue did not improve the situation with them much.

**Comment by AnHeuermann** [comment]
> What do you think?

Sounds reasonable, but its harder to implement than I initially thought because of the way our code generation is.

**Comment by casella** [comment]
Is it doable in a reasonable amount of time? We don't need anything particularly sophisticated.

**Comment by AnHeuermann** [comment]
> Is it doable in a reasonable amount of time? We don't need anything particularly sophisticated.

I don't think so. The structure inside template functions `daeExpCall` and `daeExp` is ... challenging.
We need to add a function call to the temp variable generated by `noEvent(exp)`. We have functionalities to add additional variables and expressions in front of `exp`, but not after `exp` is calculated. And we don't have access to the name of the temp variable(s) generated for `exp`.

**Comment by casella** [comment]
> So we need to follow @kabdelhak's idea to generate zero-crossing-like equations and log them all at once after the ODE-right-hand side is evaluated or change large parts of `daeExp`.

Sounds good to me, as long as the logging is done every time the RHS is evaluated, not only at the end of the accepted simulation step. Otherwise we could be missing a lot of stuff going wrong during Newton iterations, Jacobian computations, etc.
